### DISCUSSION TITLE
Deleting projects

### DISCUSSION THREAD

**Discussion by mgryning** [VsixCommunity/Community.VisualStudio.Toolkit#250] [Q&A] [answered]
Hi.

Is there a way to delete a project from the solution using the toolkit?

For instance, the below code is using a mix of the toolkit and the "old" DTE object, but the conversion is invalid:

var project = await ItemHelper.FindProjectAsync("api"); //use VS...
var solution = await DTEItemHelper.GetSolutionAsync(); //Get the DTE80 object
solution.Remove((EnvDTE.Project)project); //To use DTE80, i also have to have project in the right format

It would be helpful to have this functionality in the toolkit, mainly for convenience, but also because it seems like there exist 3-4 ways to remove a project, and it's difficult to estimate which one to use. If it's not planned to be added, is there a way I can remove the project via a mix of the toolkit and existing functionality?

Br.

**Comment by reduckted** [comment, marked as answer]
[IVsSolution.CloseSolutionElement](https://docs.microsoft.com/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution.closesolutionelement) is the method that you need to use.

> To close the entire solution, call `CloseSolutionElement(grfCloseOpts, null, null)`. 
> To close a single project, call `CloseSolutionElement(grfCloseOpts, phier, null)`. 
> To close a single document, call `CloseSolutionElement(grfCloseOpts, null, docCookie)`.

Here's an example using the toolkit:
```cs
Project? activeProject = await VS.Solutions.GetActiveProjectAsync();

if (activeProject != null) {
    IVsSolution solution = await VS.Services.GetSolutionAsync();
    activeProject.GetItemInfo(out IVsHierarchy hierarchy, out _, out _);
    solution.CloseSolutionElement(0, hierarchy, 0);
}
```

**Comment by mgryning** [comment]
Great, thank you! I would never have guessed that it was called CloseSolution :-)

**Comment by madskristensen** [comment]
Perhaps we should add a more friendly way to do this from the toolkit's `Solution` object. Something like `mySolution.RemoveProject(theProject)`;
